### PR TITLE
Phase2 tracker DQM package

### PR DESCRIPTION
**Pull request by gsaha009**
#### PR description:
In this PR, two sub packages are created for Phase2 tracker DQM and Validation. 
1.  Creating DQM/SiTrackerPhase2 and Validation/SiTrackerPhase2V packages.
2.  The main source code already existed in SimTracker/SiPhase2Digitizer package and was used in a standalone mode.
3.  Changes in DQM step: moved Phase2TrackerMonitorDigi module from SimTracker/SiPhase2Digitizer/test to DQM/SiTrackerPhase2/plugins. Updated DQM offline configs and autoDQM config to include this module in standard phase2 workflow.
4.  Changes in Validation step: moved Phase2TrackerValidateDigi module from SimTracker/SiPhase2Digitizer/test to Validation/SiTrackerPhase2V/plugins. Updated Validation configs and autoValidation config to include this module in standard phase2 workflow.
5.  Config files for 3, 4 are named following central DQM recommendations ([twiki](https://twiki.cern.ch/twiki/bin/viewauth/CMS/DQMOfflineGuide))

#### PR validation:
Tests done with workflows for phase2 : 20034.0, 21234.0, 22834.0, 23234.0

### CODE REVIEW

**Comment by mmusich** [comment]
@emiglior FYI

**Comment by emiglior** [comment]
adding @pwittich

**Comment by fabiocos** [comment]
please test

**Comment by jfernan2** [comment]
Creating the package in cms-bot
https://github.com/cms-sw/cms-bot/pull/1225

**Comment by jfernan2** [comment]
This PR also needs:
https://github.com/cms-sw/cms-bot/pull/1224

**Comment by jfernan2** [comment]
@gsaha009 please modify header description of this PR to avoid misunderstandings: this PR introduces DQM/SiTrackerPhase package not DQM/TrackerPhase2
Thanks.
> 1. Creating DQM/TrackerPhase2 and Validation/SiTrackerPhase2V packages.
> 3. Changes in DQM step: moved Phase2TrackerMonitorDigi module from SimTracker/SiPhase2Digitizer/test to DQM/TrackerPhase2/plugins. Updated DQM offline configs and autoDQM config to include this module in standard phase2 workflow.

**Comment by jfernan2** [comment]
please test 22834.0, 23234.0

**Comment by jfernan2** [comment]
please test workflow 22834.0, 23234.0

**Comment by jfernan2** [comment]
+1
Changes as expected in Phase wf

**Comment by kpedro88** [comment]
+upgrade
moving code from digitizer package to validation package

**Comment by civanch** [comment]
+1

File: Validation/SiTrackerPhase2V/plugins/BuildFile.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+<use   name="SimDataFormats/TrackerDigiSimLink"/>
+<use   name="DataFormats/SiPixelDigi"/>
+<use   name="DataFormats/SiPixelDetId"/>
+<use   name="DataFormats/DetId"/>
+<use   name="Geometry/Records"/>
+<use   name="Geometry/CommonDetUnit"/>
+<use   name="Geometry/TrackerGeometryBuilder"/>
+<use   name="DQMServices/Core"/>
+<use   name="DataFormats/Common"/>
+<use   name="SimTracker/SiPhase2Digitizer"/>
```
Comment (fabiocos) [inline Validation/SiTrackerPhase2V/plugins/BuildFile.xml]: @gsaha009 why is the plugin declaration missing?

File: DQM/SiTrackerPhase2/plugins/BuildFile.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+<use   name="SimDataFormats/TrackerDigiSimLink"/>
+<use   name="DataFormats/SiPixelDigi"/>
+<use   name="DataFormats/SiPixelDetId"/>
+<use   name="DataFormats/DetId"/>
+<use   name="Geometry/Records"/>
+<use   name="Geometry/CommonDetUnit"/>
+<use   name="Geometry/TrackerGeometryBuilder"/>
+<use   name="DQMServices/Core"/>
+<use   name="DataFormats/Common"/>
+<use   name="SimTracker/SiPhase2Digitizer"/>
```
Comment (fabiocos) [inline DQM/SiTrackerPhase2/plugins/BuildFile.xml]: @gsaha009 why is the plugin declaration missing?

**Review by fabiocos — COMMENTED** [review]
@gsaha009 could you please verify these Buildfiles with the ```createBuildFile.pl``` utility?

File: DQM/SiTrackerPhase2/plugins/BuildFile.xml (hunk 2 above)
Comment (gsaha009) [reply DQM/SiTrackerPhase2/plugins/BuildFile.xml]: I have checked and updated the buildfiles. Please ckeck

**Comment by fabiocos** [comment]
@gsaha009 were the performances of this code verified? I run test wf 22024.0 with 4 threads on the usual 10 events and I see

```
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 5  vsize = 6485.36 deltaVsize = 0 rss = 3802.88 delta = 0
[4] run: 1 lumi: 1 event: 4  vsize = 6613.36 deltaVsize = 128 rss = 3903.3 delta = 100.418
[5] run: 1 lumi: 1 event: 3  vsize = 6709.36 deltaVsize = 96 rss = 3979.54 delta = 76.2422
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[7] run: 1 lumi: 1 event: 10  vsize = 6709.36 deltaVsize = 0 rss = 4025.81 delta = 46.2656
[6] run: 1 lumi: 1 event: 7  vsize = 6709.36 deltaVsize = 0 rss = 3982 delta = 2.45703
[5] run: 1 lumi: 1 event: 3  vsize = 6709.36 deltaVsize = 96 rss = 3979.54 delta = 76.2422
```

to be compared with the base

```
[3] run: 1 lumi: 1 event: 1  vsize = 6739.21 deltaVsize = 32 rss = 3559.29 delta = 78.8008
[2] run: 1 lumi: 1 event: 2  vsize = 6707.21 deltaVsize = 64 rss = 3480.49 delta = 54.4258
[7] run: 1 lumi: 1 event: 9  vsize = 6839.21 deltaVsize = 100 rss = 3669.88 delta = 189.387
[8] run: 1 lumi: 1 event: 10  vsize = 6903.21 deltaVsize = 64 rss = 3663.71 delta = -6.17188
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[10] run: 1 lumi: 1 event: 6  vsize = 6903.21 deltaVsize = 0 rss = 3563.12 delta = -100.586
[9] run: 1 lumi: 1 event: 8  vsize = 6903.21 deltaVsize = 0 rss = 3654.51 delta = -9.19531
[8] run: 1 lumi: 1 event: 10  vsize = 6903.21 deltaVsize = 64 rss = 3663.71 delta = -6.17188
```

The RSS increase is of the order of 8%. The increase in the total DQM memory usage is also non negligible according to ```dqmMemoryStats```:

```
< 151719.35 KiB    TrackerPhase2ITDigi/DigiMonitor
13d11
< 5431.48 KiB      TrackerPhase2OTDigi/DigiMonitor
71c69
< Total bytes: 867914.04 KiB
---
> Total bytes: 710707.47 KiB
```

**Comment by fabiocos** [comment]
@gsaha009 the new packages basically contain a single plugin each, do you expect a further expansion?

**Comment by fabiocos** [comment]
please test workflow 22834.0, 23234.0

**Comment by mmusich** [comment]
@fabiocos

> the new packages basically contain a single plugin each, do you expect a further expansion?

much. This is the first stub of the still undeveloped phase2 tracker DQM.

**Comment by fabiocos** [comment]
@jfernan2 @civanch @kpedro88 the update is purely technical, I will consider your previous signatures as valid. Please keep in mind that this update will not be innocent in terms of performances

**Comment by jfernan2** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Comment by fabiocos** [comment]
@gsaha009 @mmusich I see there is a problem in the IB:

```
----- Begin Fatal Exception 26-Nov-2019 18:28:11 CET-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 2 event: 105 stream: 3
   [1] Calling method for module Phase2TrackerMonitorDigi/'pixDigiMon'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::DetSetVector&lt;PixelDigi&gt;
Looking for module label: simSiPixelDigis
Looking for productInstanceName: Pixel

----- End Fatal Exception -------------------------------------------------
```

**Comment by mmusich** [comment]
@fabiocos as far as I see it's the only affected wf (a bit peculiar). Is it the only premixed wf in the upgrade matrix?

File: DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py [hunk 3]
```diff
@@ -11,88 +11,124 @@
...
         Nxbins = cms.int32(3000),
         xmin   = cms.double(-3000.),
         xmax   = cms.double(3000.),
         Nybins = cms.int32(1250),
         ymin   = cms.double(0.),
-        ymax   = cms.double(1250.)
+        ymax   = cms.double(1250.),
+        switch = cms.bool(False)
     )
 )
```
Comment (mmusich) [inline DQM/SiTrackerPhase2/python/Phase2TrackerMonitorDigi_cfi.py]: @gsaha009.
I think that in this file `InnerPixelDigiSource` should be customized in the case of premixing.
Something like:
```
from Configuration.ProcessModifiers.premix_stage2_cff import premix_stage2
(premix_stage2 & phase2_tracker).toModify(digiMon,
    InnerPixelDigiSource = "mixData:Pixel"
)
```
or something similar (see e.g. https://cmssdt.cern.ch/dxr/CMSSW/source/RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py#31)

**Comment by mmusich** [comment]
@fabiocos issue is addressed here: https://github.com/cms-sw/cmssw/pull/28486/files

**Pull request merged**